service/lib/dinstaller/storage/proposal.rb

```diff
@@ -76,56 +78,60 @@ def device_label(device)
+      # exist or not in the real context of the proposal and its settings.
       #
-      # @return [Array<String>]
-      def candidate_devices
-        raise NoProposalError unless proposal
-
-        proposal.settings.candidate_devices
+      # @return [Array<Volumes>]
+      def volume_templates
+        VolumesGenerator.new(default_specs).volumes
       end
 
-      # Whether the proposal should create LVM devices
+      # Volumes from the specs used during the calculation of the storage proposal
       #
-      # @raise [NoProposalError] if no proposal yet
+      # Not to be confused with settings.volumes, which are used as starting point for creating the
+      # volume specs for the storage proposal.
       #
-      # @return [Boolean]
-      def lvm?
-        raise NoProposalError unless proposal
+      # @return [Array<Volumes>]
+      def calculated_volumes
+        return [] unless proposal
+
+        generator = VolumesGenerator.new(specs_from_proposal,
+          planned_devices: proposal.planned_devices)
+        volumes = generator.volumes(only_proposed: true)
 
-        proposal.settings.use_lvm
+        # FIXME: setting this should be a responsibility of VolumesGenerator or any other component,
+        # but this is good enough until we implement fine-grained control on encryption
+        volumes.each { |v| v.encrypted = proposal.settings.use_encryption }
+
+        volumes
       end
 
       # Calculates a new proposal
       #
-      # @param settings [Hash] settings to calculate the proposal
-      #   (e.g., { "use_lvm" => true, "candidate_devices" => ["/dev/sda"]}). Note that keys should
-      #   match with a public setter.
-      #
+      # @param settings [ProposalSettings] settings to calculate the proposal
       # @return [Boolean] whether the proposal was correctly calculated
-      def calculate(settings = {})
-        proposal_settings = generate_proposal_settings(settings)
+      def calculate(settings = nil)
+        @settings = settings || ProposalSettings.new
+        @settings.freeze
+        proposal_settings = to_y2storage_settings(@settings)
 
-        @proposal = Y2Storage::GuidedProposal.initial(
-          settings:      proposal_settings,
-          devicegraph:   probed_devicegraph,
-          disk_analyzer: disk_analyzer
-        )
-        save
-        changed!
+        @proposal = new_proposal(proposal_settings)
+        storage_manager.proposal = proposal
+
+        @on_calculate_callbacks.each(&:call)
 
         !proposal.failed?
       end
 
       # Storage actions manager
       #
-      # @fixme this method should directly return the actions
+      # FIXME: this method should directly return the actions
       #
       # @return [Storage::Actions]
       def actions
-        # FIXME: this class could receive the storage manager instance
+        # FIXME: this class could receive the actiongraph associated to the current proposal
```

Comment:
Yes, please, it would make it easier to understand the role of the Actions class without having to check the code.